worker/src/socket.rs

```diff
@@ -0,0 +1,340 @@
+    reader: worker_sys::ReadableStreamDefaultReader,
+    write: Option<Writing>,
+    read: Option<Reading>,
+    close: Option<Closing>,
+}
+
+impl Socket {
+    fn new(inner: worker_sys::Socket) -> Self {
+        let _writable = inner.writable();
+        let writer = _writable.get_writer();
+        let _readable = inner.readable();
+        let reader = _readable.get_reader();
+        Socket {
+            inner,
+            _writable,
+            writer,
+            _readable,
+            reader,
+            read: None,
+            write: None,
+            close: None,
+        }
+    }
+
+    /// Closes the TCP socket. Both the readable and writable streams are forcibly closed.
+    pub async fn close(&mut self) -> Result<()> {
+        JsFuture::from(self.inner.close()).await?;
+        Ok(())
+    }
+
+    /// This Future is resolved when the socket is closed
+    /// and is rejected if the socket encounters an error.
+    pub async fn closed(&self) -> Result<()> {
+        JsFuture::from(self.inner.closed()).await?;
+        Ok(())
+    }
+
+    /// Upgrades an insecure socket to a secure one that uses TLS,
+    /// returning a new Socket. Note that in order to call this method,
+    /// you must set [`secure_transport`](SocketOptions::secure_transport)
+    /// to [`StartTls`](SecureTransport::StartTls) when initially
+    /// calling [`connect`](connect) to create the socket.
+    pub async fn start_tls(self) -> Socket {
+        let inner = self.inner.start_tls();
+        Socket::new(inner)
+    }
+
+    pub fn builder() -> ConnectionBuilder {
+        ConnectionBuilder::default()
+    }
+}
+
+fn js_value_to_std_io_error(value: JsValue) -> std::io::Error {
+    let s = if value.is_string() {
+        value.as_string().unwrap()
+    } else if let Some(value) = value.dyn_ref::<js_sys::Error>() {
+        value.to_string().into()
+    } else {
+        format!("Error interpreting JsError: {:?}", value)
+    };
+    std::io::Error::new(std::io::ErrorKind::Other, s)
+}
+impl tokio::io::AsyncRead for Socket {
+    fn poll_read(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> std::task::Poll<std::io::Result<()>> {
+        // Writes as much as possible to buf, and stores the rest in internal buffer
+        fn handle_data(
+            buf: &mut tokio::io::ReadBuf<'_>,
+            data: Vec<u8>,
+        ) -> (Reading, Poll<std::io::Result<()>>) {
+            let idx = buf.remaining().min(data.len());
+            console_log!("Ready {}/{} bytes", data.len(), idx);
+            buf.put_slice(&data[..idx]);
+            if idx == data.len() {
+                console_log!("Read to end");
+                (Reading::None, Poll::Ready(Ok(())))
+            } else {
+                console_log!("Storing {} in internal buffer", data.len() - idx);
+                let text = std::str::from_utf8(&data[idx..]).unwrap();
```

Comment:
This assumes UTF8. Do we want to leave that assumption to the client code and let it handle the decoding? (Similar to how it's done in JS with TextDecoder)

Reply:
This was only used for debugging, I will not be doing this in practice.